### PR TITLE
Improved navigation and permissions

### PR DESCRIPTION
**Pull request by philips77**
This PR contains redesigned implementation of Navigation module and improvements in Permissions module.

## Navigation

The *navigation* module has been redesigned with the same principles in mind:
1. Navigation implemented as `NavigationView`.
2. Each navigation destination is identified by `DestinationId`.
3. Each destination defines how it should be rendered. These definitions can be in separate modules.
4. Destinations can navigate between passing arguments (->) and results (<-).

### Issues in the old implementation

1. Before, each destination could navigate to another one by giving its `DestinationId`.
The issue was, that sometimes destinations were defined in separate modules and were independent from each other.
2. Arguments and results were passed using a `ViewModel`.
The issue was, that although the `route` (ID) of the destination was saved by `NavHostController` in `SavedStateHandle` and recreated after the app was killed, the arguments and results were not.

### New implementation

> Note:
   In #7 I tried to fix the first issue by introducing a `Router` callback, which would move the modules interconnections to the main module, but that ended up being an overkill, which made navigation less readable.

1. As before, each destination is identified by `DestinationId`. A new helper method is added to make it easier:
```kotlin
val Main = createSimpleDestination("main")
``` 
If a destination requires an argument or returns results, the types have to be provided in the declaration:
```kotlin
val ColorPicker = createDestination<Float, Color>("picker")
``` 
The returned type must be supported by `writeValue` of `Parcalable`, aka `@RawValue`.

2. A module may define a destination:
```kotlin
private val ColorPickerDestination = defineDestination(ColorPicker) {
   ColorPickerScreen()
}
```
3. The arguments passed using `navigator.navigateTo(destination, args)` are saved in `SavedStateHandle` of the target destination and are preserved in configuration changes, or when the app is killed by the system to release resources. They are also available in `ViewModel`s created in the target destination when Hilt dependency injection is used, using [`SavedStateHandle`](https://developer.android.com/topic/libraries/architecture/viewmodel-savedstate).
4. Each destination may inject `Navigator` object in the view models. This object allows to navigate to another destination passing an argument.
```kotlin
@HiltViewModel
class MainViewModel @Inject constructor(
   navigator: Navigator,
   savedStateHandle: SavedStateHandle,
): ViewModel() {
   // Initialize the state from saved state.
   val color = savedStateHandle.getStateFlow<Color?>("color", null)

   init {
      // Register for navigation results in init. Results will be delivered also when the Activity was just created
      // but the color was selected before.
      navigator.resultFrom(ColorPicker)
         // Filter out results of cancelled navigation.
         .filter { it != null }
         // Save the result in SavedStateHandle.
         .onEach { savedStateHandle["color"] = it }
         // And finally, launch the flow in the ViewModelScope.
         .launchIn(viewModelScope)
    }

   fun chooseColor(lightness: Float) {
      // Navigate by providing the destination and optional argument. The argument type is defined in the destination.
      navigator.navigate(ColorPicker, lightness)
   }
}

@HiltViewModel
class ColorPickerViewModel @Inject constructor(
   navigator: Navigator,
   savedStateHandle: SavedStateHandle,
): ViewModel() {
   // Obtain the argument. The type is defined in the ColorPicker destination.
   val lightness = savedStateHandle.get(ColorPicker) ?: 1.0f

   fun returnColor(color: Int) {
      navigator.navigateUpWithResult(from = ColorPicker, result = color)
   }
}
```

## Permissions

These are the main changes:
1. The permissions module was refactored. `RequireBluetooth(scanning: Bool, ...)` was split into 2:
   * `RequireBluetooth`
   * `RequireLocation` - this is required for scanning for Bluetooth LE devices. On Android 12+ it always assumes the location is granted, as it is no longer required for scanning. At least for most of the devices (it is needed for iBeacons and Eddystone).
2. It is now possible to call `RequireBluetooth` and disabling displaying default "No Bluetooth" views by setting `contentWithoutBluetooth` parameter to an empty composable. by default it will show `NoBluetoothView`.
3. The same is true for `RequireLocation` and `RequireInternet`.
4. The files in the module were moved around, the internal views were made internal.

### CODE REVIEW

**Comment by philips77** [comment]
Things that could be improved:
- [ ] A simple ViewModel which would only handle navigation, for simple cases.
- [ ] A base ViewModel that could consume navigation events, e.g. back navigation.

File: navigation/src/main/java/no/nordicsemi/android/common/navigation/internal/NavController+Bundle.kt [hunk 1]
```diff
@@ -29,23 +29,29 @@
  * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
```
[the remark this answers is not in the document]
Comment (philips77) [reply navigation/src/main/java/no/nordicsemi/android/common/navigation/internal/NavController+Bundle.kt:30]: I did not and what are you going to do to me?

**Comment by philips77** [comment]
The latest changes added fixed navigation events and type checking for destination arguments.

**Review by sylwester-zielinski — APPROVED** [review]

File: navigation/src/main/java/no/nordicsemi/android/common/navigation/internal/NavigationViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        navigationManager.executor = this
+    }
+
+    /**
+     * The given [SavedStateHandle] will be used to store the navigation result.
+     *
+     * This has to be the instance given in navigation composable, as each ViewModel may received
+     * a different instance of [SavedStateHandle] using Hilt injections.
+     */
+    fun use(savedStateHandle: SavedStateHandle) = this
```
[the remark this answers is not in the document]
Comment (philips77) [reply navigation/src/main/java/no/nordicsemi/android/common/navigation/internal/NavigationViewModel.kt:33]: Worth it!

**Pull request merged**